Fixed #31949 -- Made builtin decorator async compatible

LomaxOnTheRun opened the pull request:
Fix for [#31949](https://code.djangoproject.com/ticket/31949).

Makes builtin decorators async compatible so that they can be applied directly to async views without having to be wrapped in additional `@async_to_sync` and `@sync_to_async` decorators.

The two main changes are the creation of `sync_async_wrapper` and the update to `make_middleware_decorator` which allow for most of the decorators to be simply converted to having a sync and an async implementation. The `sync_async_wrapper` works by offering "hooks" where calling code can pass in functions which affect the request and the response, in a cut down but very similar way to how `make_middleware_decorator` currently works.

Decorators:
- [x] `@cache_page` (uses `decorator_from_middleware_with_args`)
- [x] `@cache_control`
- [x] `@never_cache`
- [x] `@xframe_options_deny`
- [x] `@xframe_options_sameorigin`
- [x] `@xframe_options_exempt`
- [x] `no_append_slash`
- [x] `@csrf_protect` (uses `decorator_from_middleware`)
- [x] `@requires_csrf_token` (uses `decorator_from_middleware`)
- [x] `@ensure_csrf_cookie` (uses `decorator_from_middleware`)
- [x] `@csrf_exempt`
- [x] `@sensitive_variables`
- [x] `@sensitive_post_parameters`
- [x] `@gzip_page` (uses `decorator_from_middleware`)
- [x] `conditional_page` (uses `decorator_from_middleware`)
- [x] `@require_http_methods`
- [x] `@require_GET`
- [x] `@require_POST`
- [x] `@require_safe`
- [x] `@condition`
- [x] `@etag`
- [x] `@last_modified`
- [x] `@vary_on_headers`
- [x] `@vary_on_cookie`
- [x] `@user_passes_test`
- [x] `@login_required`
- [x] `@permission_required`

hendrikfrentrup commented:
Great to see work on this - I recently stumbled upon problems with the HTTP decorators and tried a similar fix (https://github.com/django/django/compare/master...hendrikfrentrup:master). As far as I can see you haven't gotten to them yet, so I can give these a go if you don't mind. I haven't looked into async testing yet, so cannot comment much on your questions.

LomaxOnTheRun commented:
@hendrikfrentrup if you want to wire up the `http` decorator please be my guest :slightly_smiling_face: I've pushed up my latest changes, I'll avoid force pushing from here onwards to avoid overwiting any of your stuff :slightly_smiling_face:

hendrikfrentrup commented:
OK, I slightly rewrote my fix to adjust to the syntax suggested by felixxm. Also, it wasn't entirely straight forward to write an async test because the test suite runs slightly different to how an application routes to a view via the URL paths (the view is not awaited in the URL path while in the test `await my_safe_view(request)` would await for the return of the HttpResponse)

Anyway, @LomaxOnTheRun if you add me as a contributor to your forked repo, I can push to this PR. Or do you want me to push to my own forked repo?

hendrikfrentrup commented:
For visibility, I've pushed the changes to an equivalent branch in my forked repo [here](https://github.com/hendrikfrentrup/django/commit/255ac6e55c0d5b8105e7823b7e93c07a7d641bf2) - I can push the same changes to this PR or merge it onto your branch. Not sure which one is more convenient.

LomaxOnTheRun commented:
@hendrikfrentrup I've sent you an invite to be a collaborator on my repo's fork. It's probably best to have everything on the same branch so that it can all be checked at the same time, given it's all similar / connected functionality.

LomaxOnTheRun commented:
@hendrikfrentrup as a heads up I've tweaked the base `sync_async_wrapper` in my last commit so if you need to change it (which it looks like you may need to to allow for changes to the request, not just the response), it might be worth pulling the branch again :slightly_smiling_face:

hendrikfrentrup commented:
Ok, I've pushed the changes - as far as I can tell your `sync_async_wrapper` isn't suitable as is for the `require_http_methods` decorator because it doesn't process the respone at all, it simply checks the request and returns a straightforward Not-Allowed Response if the method isn't allowed. It should come in handy for the `condition` based decorators though (started looking into that a bit already).

So, the test directly probes whether the decorator is sync and async capable as suggested [here](https://docs.djangoproject.com/en/3.1/topics/http/middleware/#asynchronous-support) and also direcly probes an async view being a coroutine and a sync view staying synchronous. I think that is sufficient because the functionality is essentially tested by the existing test already.

It does in deed require the provided patch to make the test pass. If I had just changed the decorated view(`my_safe_view`) to async, it doesn't actually throw an exception, just a warning which has been debated [here](https://code.djangoproject.com/ticket/31407)

hendrikfrentrup commented:
Also, @LomaxOnTheRun I've also gave some thoughts to your point #2 above. I think it should suffice to write an async test for the async version (in a way it actually would test the code twice). Also, the way I see it, any sync or async runtimes can probe whether functions are sync and/or async capable and the respective view function are decorated according to whether the view itself is sync/async. That's also why I decorated the decorator with `@sync_and_async_middleware` (I guess I am interpreting this as middleware, in this case builtin middleware). Is there a reason why your decorators aren't decorated sync/async compatible?

LomaxOnTheRun commented:
@hendrikfrentrup thanks for all of that info, I'm actually finishing up at my current job this week so I might not have time to look at this until the weekend, but I'll ping you a soon as I do.

The lack of `@sync_and_async_middleware` decorators was just because I wanted to get a draft up quickly so others could view and comment on it, but I'll need to circle back to them and add them. Thanks for the reminder 😁

hendrikfrentrup commented:
Yup, sounds good - have a good week :) I've just force pushed some minor amends hoping you have not pulled the commit. Need to get in the habit of running `isort` and `flake8` in addition to the test suite before pushing. Green check is back now though.

LomaxOnTheRun commented:
@hendrikfrentrup just a heads up that I've changed the remaining decorators (and `make_middleware_decorator`), added the documentation updates, changed the PR to `Ready for checkin` in the Django ticket system.

I've also squashed all the commits into one and added you as a co-author (which you can see in the `Commits` tab) and added you to the `AUTHORS` page, along with your Github email address (I did the same with mine). Let me know if you'd like me to change either your name or email address, or remove one or both of them and I'll happily do that :slightly_smiling_face: 

Otherwise just waiting for the reviews to start rolling in :smile:

hendrikfrentrup commented:
@LomaxOnTheRun Nice one! Thanks for adding the author notes 🙂. Also looking forward to the comments.

Hard to wrap my head around the `make_middleware_decorator`, but looks good to me. While it seems like the methods `check_request_method` and `add_response_headers` could also be re-written as class-based Middleware using the `process_request` and `process_response` approach, it's not like much would be gained from it - I find it more readable as is (prefer functional over object-oriented personally). Plus, it is closer to the current implementation.

auvipy commented:
@andrewgodwin might be of your interest

auvipy commented on docs/topics/async.txt:
```diff
@@ -116,6 +116,30 @@ mode if you have asynchronous code in your project.
 
+Async view decorators
+=====================
+
```
maybe version added annotation needed and release notes

LomaxOnTheRun replied:
What would this come under? `What's new in Django 3.X`? `Minor features`?

jacobtylerwalls replied:
I would just use your best judgment. Be sure to edit `4.0.txt`.

roysmith commented on django/utils/decorators.py:
```diff
@@ -111,13 +112,41 @@ def decorator_from_middleware(middleware_class):
     return make_middleware_decorator(middleware_class)()
 
 
+def sync_and_async_middleware(func):
```
I'm confused by the definitions of these three functions here.  From [the docs](https://docs.djangoproject.com/en/3.1/topics/http/middleware/), it looks like these are already in 3.1.

roysmith commented on tests/decorators/tests.py:
```diff
@@ -466,17 +604,583 @@ def a_view(request):
+        return '"abc123"'
+
+    def last_modified_func(self, request, *args, **kwargs):
+        return datetime(2020, 1, 1, 0, 0, 0)
```
I think `datetime(2020, 1, 1)` is the same thing, i.e. the rest of the zeros default.

roysmith reviewed (changes requested):
I also verified (debian-10,  python-3.7.3),  that the docs build, flake8 produces no errors, and all unit tests pass.

rednaks commented:
`login_required` ?

LomaxOnTheRun commented on tests/decorators/tests.py:
```diff
@@ -466,17 +604,583 @@ def a_view(request):
-        r = XFrameOptionsMiddleware(a_view)(req)
-        self.assertIsNone(r.get('X-Frame-Options', None))
+        resp = XFrameOptionsMiddleware(a_view)(req)
+        self.assertIsNone(resp.get('X-Frame-Options', None))
```
Why should this be `assertFalse`? It seems like a positive thing to keep the assertion closer to the actual value returned, and I think the only falsey value returned here is `None`.

LomaxOnTheRun commented on tests/decorators/tests.py:
```diff
@@ -466,17 +604,583 @@ def a_view(request):
+        async def an_async_view(request):
+            return HttpResponse()
+        resp = await an_async_view(HttpRequest())
+        cache_control_headers = resp.headers['Cache-Control'].split(', ')
```
I've gone with `cache_control_items` to mirror Python's dictionary `.items()` functionality as everything returned is either an explicit key/value pair, or an implicit boolean one.

LomaxOnTheRun commented on tests/decorators/tests.py:
```diff
@@ -466,17 +604,583 @@ def a_view(request):
+            return HttpResponse()
+        resp = a_view(HttpRequest())
+        self.assertEqual(resp.status_code, 200)
+        self.assertIn('Header, Another-header', resp.headers['Vary'])
```
I think it might actually keep the order because it's only every using a set to check new values against existing ones, but it keeps track of (and then recombines) values all within lists. I've changed it anyway since it actually makes for a slightly stronger test.

LomaxOnTheRun reviewed (commented):
@roysmith thank you for all of your reviews, I've either changed things accordingly, or pushed back with an explanation. If you feel like any of the points merit further discussion I'm very happy to discus them further :slightly_smiling_face: 

I've also not added release notes yet since I doubt that it'll be shipped with version 3.2.1 which is due out tomorrow, so I'll add it in for (I imagine) 3.2.2, once that document has been created in the repo.

auvipy commented:
its a major change so it have to be targetted for v4.0

roysmith commented:
I didn't see anything that was broken, these were all just style and refactoring suggestions, so I'm good with whichever ones you decided to pick up or reject.  Thanks for working on the async support; this will be a nice step forward in django's feature set.

> On May 3, 2021, at 4:53 PM, Ben Lomax ***@***.***> wrote:
> 
> @LomaxOnTheRun commented on this pull request.
> 
> @roysmith <https://github.com/roysmith> thank you for all of your reviews, I've either changed things accordingly, or pushed back with an explanation. If you feel like any of the points merit further discussion I'm very happy to discus them further 🙂
> 
> I've also not added release notes yet since I doubt that it'll be shipped with version 3.2.1 which is due out tomorrow, so I'll add it in for (I imagine) 3.2.2, once that document has been created in the repo.
>

jacobtylerwalls commented on docs/topics/async.txt:
```diff
@@ -110,6 +110,32 @@ mode if you have asynchronous code in your project.
+Async view decorators
+=====================
+
+.. versionadded:: 3.2.1
```
```suggestion
.. versionadded:: 4.0
```

jacobtylerwalls commented on docs/topics/async.txt:
```diff
@@ -110,6 +110,32 @@ mode if you have asynchronous code in your project.
+    def my_sync_view(request):
+        pass
+
+    # ... or it can be asyncronous
```
async**h**ronous

auvipy reviewed (changes requested):
please rebase this, I believe this could be  good for v 4.0. also did you considered sync_to_async from asgiref?

carltongibson commented on django/utils/decorators.py:
```diff
@@ -147,33 +176,75 @@ def callback(response):
 
 
-def sync_only_middleware(func):
+def sync_async_wrapper(
+    view_func, process_request=None, process_response=None, **decorator_kwargs
+):
```
Not immediately sure about this function... 🤔

Naming for one... (we can ponder that). 

But more, it's half-encapsulating the pattern we see quite a few times in this PR — but then it's not fully applied. SO... either it's needs abstracting a little so it does apply OR it's not paying the rent, and we should inline it. 
(To think about.)

carltongibson reviewed (commented):
Hi @LomaxOnTheRun — Thanks for this. 

Sorry for the slow review — I've looked at it a few times, but needed some space to really get my head into it. 

I've left some comments, which fall into a few general areas: 

* Can you **not** make unrelated changes in the PR. If you think they should be made you can suggest them separately, but they hide what's really going on. 
* We should find a way to test the decorators with both `def` and `async def` views using `subTest()` if possible — there's a lot of identical tests there. 
* The are preliminary refactoring changes that can be made separately to help reduce the footprint again. They can be made in small commits/PRs that are easier to assess. 
* Then, I think we wouldn't be far off the core pattern. I'll have a ponder on that...

Are you still up for working on this? 

Thanks again!

carltongibson commented:
Hey @LomaxOnTheRun — great, thanks! Just FYI, I'm trying to pick up and process a bunch of these async tickets (targeting Django 4.2) so this is on my list for that. 👍

Thanks for the input! 🎁

LomaxOnTheRun commented:
@auvipy I'm looking at getting this rebased (and updated as per the latest code review). As for:

> also did you considered sync_to_async from asgiref?

The [Django ticket](https://code.djangoproject.com/ticket/31949) which prompted this work is about not having to use `sync_to_async` and `async_to_sync` to switch between async and sync decorators, and to allow the builtin view decorators to just be able to handle both. If you meant using `sync_to_async` in the code, it seems to me that a direct implementation is better than relying on a third party library when we're able to implement it directly :slightly_smiling_face:

LomaxOnTheRun commented:
@carltongibson just to let you know that I'm working my way through all of the decorators again one by one, as it's been 14 months since I last worked on this branch and things on `main` have changed since then. I'm using [this branch](https://github.com/LomaxOnTheRun/django/pull/1) on my personal repo to go through and makes changes and notes as I go, so I can have a scratchpad while keeping all of your comments on this branch as they are.

From my initial foray, it looks like the decorators split into roughly 3 piles:

1. Easy to refactor decorators.
2. Decorators that use `decorator_from_middleware`(`_with_args`)
    - These don't currently work on my branch but it seems that once I figure out how to make one work the rest should be easy
3. Other misc awkward ones for various reasons
    - These are more of a pain as they will likely need attention given to each one individually

My plan is to get at least pile 1 done in this PR, and look at possibly splitting piles 2 and 3 into separate PRs. Let me know if that seems like a good / bad idea, and also feel free to move the conversation to the other branch (fair warning, it's quite fluid and prone to getting rebased) :slightly_smiling_face:

carltongibson commented:
@LomaxOnTheRun — great. Plug away until you feel like your progress is slowing, and then I can have another look. Thanks for the input!

LomaxOnTheRun commented:
@carltongibson I think I'm at a good point to push up a PR which covers most of the decorators. Would it be better to rebase this one (to keep the full history of comments) or to create a new PR (which would be cleaner)?

I've also got some commits which are an attempt to reduce the duplication in the decorator code, but I think they would be best shown in a separate PR, based on the main PR.

auvipy commented:
you can push here for now

LomaxOnTheRun commented on tests/decorators/tests.py:
```diff
@@ -474,6 +474,21 @@ def method(self):
+        for decorator in decorators:
+            with self.subTest(decorator):
+                self.assertTrue(decorator.sync_capable)
+                self.assertTrue(decorator.async_capable)
```
This test has been written in this way so further decorators which are made sync and async capable can be tested by being added to the `decorators` tuple, instead of creating a whole new test per decorator.

LomaxOnTheRun commented on tests/decorators/tests.py:
```diff
@@ -492,7 +493,7 @@ def test_decorators_syanc_and_async_capable(self):
 
 
-class XFrameOptionsDecoratorsTests(TestCase):
+class XFrameOptionsDecoratorsTests(SimpleTestCase):
```
This has been changed from `TestCase` to `SimpleTestCase` for two reasons:

1. `TestCase` doesn't seem to await async functions properly. I'm not sure exactly why, but this doesn't seem like the right PR o fix it in.
2. As per the [Django docs](https://docs.djangoproject.com/en/dev/topics/testing/tools/#testcase) since we're not using a database for any of these tests, we should be using `SimpleTestCase` instead of `TestCase`.

LomaxOnTheRun commented on django/utils/decorators.py:
```diff
@@ -156,7 +156,47 @@ def callback(response):
                 return response
 
-            return _wrapper_view
+            @wraps(view_func)
+            async def _wrapper_view_async(request, *args, **kwargs):
+                if hasattr(middleware, "process_request"):
+                    result = middleware.process_request(request)
+                    if result is not None:
+                        return result
+                if hasattr(middleware, "process_view"):
+                    result = middleware.process_view(request, view_func, args, kwargs)
+                    if result is not None:
+                        return result
+                try:
+                    response = await view_func(request, *args, **kwargs)
+                except Exception as e:
+                    if hasattr(middleware, "process_exception"):
+                        result = middleware.process_exception(request, e)
+                        if result is not None:
+                            return result
+                    raise
+                if hasattr(response, "render") and callable(response.render):
+                    if hasattr(middleware, "process_template_response"):
+                        response = middleware.process_template_response(
+                            request, response
+                        )
+                    # Defer running of process_response until after the template
+                    # has been rendered:
+                    if hasattr(middleware, "process_response"):
+
+                        def callback(response):
+                            return middleware.process_response(request, response)
+
+                        response.add_post_render_callback(callback)
+                else:
+                    if hasattr(middleware, "process_response"):
+                        return middleware.process_response(request, response)
+                return response
```
This duplicates almost all of the code for the `_wrapper_view_sync` function - this one is just an async function and awaits the `view_func`. I have [another PR](https://github.com/LomaxOnTheRun/django/pull/7) which looks at how we could DRY this up, as well as the repetition between the `sync_async_wrapper` and the `make_middleware_decorator` functions, but it's not a super clean refactor.

I think it offers some extra positives of sharing a code path for function decorators and class based decorators, but it feels like it should probably be done as a separate PR so a full discussion can be had for it in its own space.

LomaxOnTheRun commented:
@carltongibson @auvipy I've pushed my latest work to this branch, split out into smaller commits (which will hopefully make things a little easier to review). As per the PR description there are still 4 decorators which are not covered, but I'm open to either working on fixing those in this PR or creating a new PR specifically to address those.

I've also got a [separate draft PR](https://github.com/LomaxOnTheRun/django/pull/7) open where I investigated how we might DRY up the code both between the `sync_async_wrapper` and `make_middleware_decorator` functions, as well as the sync and async inner function for both. It gets complicated enough that I think it warrants a different PR and discussion to talk about how to best do it.

For the tests, I've not yet found any way to dry up the sync / async repetition, _especially_ since we're specifically testing functions with decorators. I'm keen to hear from anyone if they have ideas on how to approach this.

gnat commented:
Thank you for your excellent work so far @LomaxOnTheRun ! 

>As per the PR description there are still 4 decorators which are not covered, but I'm open to either working on fixing those in this PR or creating a new PR specifically to address those.

May I suggest splitting out the unfinished decorators into a seperate ticket to clear the path forward? 

Speaking as a newcomer to Django as of 4.1, I've bumped into a lack of async support for `@csrf_exempt` already while implementing an async endpoint.

I suspect there may be a lot of newcomers running into these as a blocker as people re-visit Django to try out the excellent new async ORM, and it would be ideal if this were shipped out soon. @carltongibson

LomaxOnTheRun commented on django/views/decorators/http.py:
```diff
@@ -118,25 +133,39 @@ def get_last_modified():
             return response
 
-        return inner
+        return inner_async if asyncio.iscoroutinefunction(func) else inner_sync
```
This was the neatest I could get it. The reason I can't just use `sync_async_wrapper` is that we would calculate `res_etag` and `res_last_modified` and use them in `process_request`, but then would need to access them again in the `process_response` function. To do this we would need to either:

1. Have some sort of dict / object to store them against (all the ones I could think of felt black-magic-y to me)
2. Recalculate them during `process_reponse` (but since we allow the user to pass a function, we can't guarantee that it will be idempotent and we might trigger things twice - this felt like a bad gotcha)

This is what the code partially looks like, with the problem highlighted:

```python
    def decorator(func):
        def _get_res_etag(request, *args, **kwargs):
            # The value from etag_func() could be quoted or unquoted.
            res_etag = etag_func(request, *args, **kwargs) if etag_func else None
            res_etag = quote_etag(res_etag) if res_etag is not None else None
            return res_etag

        def _get_last_modified(request, *args, **kwargs):
            # Compute values (if any) for the requested resource.
            if last_modified_func:
                dt = last_modified_func(request, *args, **kwargs)
                if dt:
                    if not timezone.is_aware(dt):
                        dt = timezone.make_aware(dt, datetime.timezone.utc)
                    return int(dt.timestamp())

        def _process_request(request, *args, **kwargs):
            res_etag = _get_res_etag(request, *args, **kwargs)
            res_last_modified = _get_last_modified(request, *args, **kwargs)
            return get_conditional_response(
                request,
                etag=res_etag,
                last_modified=res_last_modified,
            )

        def _process_response(request, response, **decorator_kwargs):

            # --- WE DON'T HAVE res_last_modified OR res_etag HERE ---

            # Set relevant headers on the response if they don't already exist
            # and if the request method is safe.
            if request.method in ("GET", "HEAD"):
                if res_last_modified and not response.has_header("Last-Modified"):
                    response.headers["Last-Modified"] = http_date(res_last_modified)
                if res_etag:
                    response.headers.setdefault("ETag", res_etag)

        return sync_async_wrapper(
            func,
            process_request=_process_request,
            _process_response=_process_response,
        )

    return decorator
```

LomaxOnTheRun commented:
@carltongibson @gnat I've actually worked out how to make the last few decorators work for sync and async views. I've also updated the documentation and the `4.2.txt` release notes, so I think this PR just needs a review now.

carltongibson commented:
Thanks @LomaxOnTheRun — you're on my list 😉

carltongibson commented:
Hi @LomaxOnTheRun — I've left some initial comments on the ticket, as I'm not sure about the approach here. 😬

I'm still thinking about this but, just looking at `@login_required` and `LoginRequiredMixin` — could we **just** wrap `@login_required` to force a sync view (if needed) and use the approach from `View.options` (linked in comment on ticket) for `LoginRequiredMixin`? (Then can we generalise from there? 🤔) — That's where I'm thinking next.

LomaxOnTheRun commented:
@carltongibson apologies for the lack of response on my part, your last comment came in right as a bunch of life stuff kicked off for me. I'm hoping to be able to get eyes on this again next week (or the week after at the latest) 🙂

I'll start with the `login_required` pieces you mentioned, although I thought your original message also included some other points that I can't see any more 🤔

carltongibson reviewed (commented):
Hi @LomaxOnTheRun — thanks for updating this. I'm thinking about it now. 

One initial comment: I'm not sure about all the new tests living in the one `decorators/tests.py` file. — As I read it that file is mostly about confirming the mechanics of the decoration, not behaviour. I'm expecting to see view tests in e.g. `auth_tests`, `cache/tests.py`, and so on. As you have it currently, I go looking for `@cache_page` behaviour and I'm having to look or separate modules. 🤔

carltongibson commented on django/views/decorators/common.py:
```diff
@@ -9,8 +13,17 @@ def no_append_slash(view_func):
     # view_func.should_append_slash = False would also work, but decorators are
     # nicer if they don't have side effects, so return a new function.
     @wraps(view_func)
-    def wrapper_view(*args, **kwargs):
+    def _wrapper_view_sync(*args, **kwargs):
         return view_func(*args, **kwargs)
 
+    @wraps(view_func)
+    async def _wrapper_view_async(*args, **kwargs):
+        return await view_func(*args, **kwargs)
+
+    if asyncio.iscoroutinefunction(view_func):
+        wrapper_view = _wrapper_view_async
+    else:
+        wrapper_view = _wrapper_view_sync
+
```
Here we're not actually consuming the response. 

I think it's worth looking at reduce the boilerplate in these cases, by simply marking the wrapper as a coroutine, [similar to in View.as_view()](https://github.com/django/django/blob/cbc0fb370583c6371eb2901f752ce286f008a6a4/django/views/generic/base.py#L118-L120). (When `wrapper_view` is called it will return the coroutine object, which will then be awaited by the calling code.)

carltongibson commented on django/views/decorators/http.py:
```diff
@@ -40,25 +42,46 @@ def inner(request, *args, **kwargs):
-            return func(request, *args, **kwargs)
 
-        return inner
+        @wraps(func)
+        def _wrapper_view_sync(request, *args, **kwargs):
+            response = _process_request(request)
+            if response is not None:
+                return response
+
+            return func(request)
+
+        @wraps(func)
+        async def _wrapper_view_async(request, *args, **kwargs):
+            response = _process_request(request)
+            if response is not None:
+                return response
+
+            return await func(request)
+
+        if asyncio.iscoroutinefunction(func):
+            return _wrapper_view_async
+        return _wrapper_view_sync
```
Again, when we're not actually consuming the wrapped view (when there's no process response) we should just be able to mark the wrapper as a coroutine, rather than duplicating it throughout.

carltongibson commented:
Closing [as per ticket](https://code.djangoproject.com/ticket/31949#comment:27)
